apps/docs/content/components/VideoPlayer.mdx

```diff
@@ -42,7 +43,24 @@ TODO: Fix asset loading in environments that don't use `pathPrefix`. This is a t
 ## Without branding
 
 ```jsx live
-<VideoPlayer branding={false} title="GitHub media player">
+<VideoPlayer hasBranding={false} title="GitHub media player">
+  <VideoPlayer.Source src="https://primer.github.io/brand/assets/example.mp4" />
+  <VideoPlayer.Track src="https://primer.github.io/brand/assets/example.vtt" />
+</VideoPlayer>
+```
+
+## With custom controls
+
+```jsx live
+<VideoPlayer
+  title="GitHub media player"
+  renderControls={() => (
+    <ControlsBar>
+      <PlayPauseButton />
+      <SeekControl />
+    </ControlsBar>
+  )}
```

Comment:
I'd usually expect these to be flags on the VideoPlayer parent like `showPlayPause` or `showControls`. Particularly as these components [don't seem to take any props](https://github.com/primer/brand/pull/634/files#diff-40d32b8a29dde73e255060785b9239f79dcef575383ee2290479a91ef83ec563R6), I don't think they need to be composable children personally 🤔, but if they did can we stick to name spacing these for consistency please? E.g. `<VideoPlayer.SeekControl />`. Could you elaborate on why you went with this approach vs the two I described above so I'm understanding this correctly.

Right now, the two issues I'm seeing with this approach are:

- We require passing an anonymous inline function inside the return render statement, so you can't take advantage of memoization. We also haven't used this pattern in PB before, so this is a precedent.
- We're surfacing sub-components in the root of the package, some of which could overlap with future component names. E.g. ControlsBar.

  ![Screenshot 2024-07-12 at 10 05 11](https://github.com/user-attachments/assets/58657b0e-eadd-4a4b-a59a-337b90896f7e)

Reply:
I went for the more composable approach as, imo, that offers a tonne of flexibility without requiring us to expand the `VideoPlayer` component's API too much now (with the `showPlayPause`, `showControls`, etc props) and defends against future use-cases we might not have considered at this point, eg attaching analytics events to the controls, modifying or adding additional controls, etc.

As with anything, I'm not rigid on this approach at all and if you really don't like it I can swap it out, but it was very much a conscious decision to take this approach and not just the first thing that came to mind. I do very much appreciate the flexibility that this pattern offers, and my experience is that it requires less maintenance in the long-run.

On your two bullet points:

> We require passing an anonymous inline function inside the return render statement, so you can't take advantage of memoization. We also haven't used this pattern in PB before, so this is a precedent.

Very true, I've not seen this explicit pattern in any other places in our codebase (we do have a couple of `renderXyz()` functions which return JSX, but none which are taken as props afaict) but I don't see this approach as particularly radical.

If someone wanted to memoize that function then they could.

```tsx
const MyComponent = () => {
  const controls = useCallback(() => (
    <ControlsBar>
      <PlayPauseButton />
      <SeekControl />
    </ControlsBar>
  ), [])

  return (
    <VideoPlayer
      renderControls={controls}
    />
  );

}
```

But to be honest, in that case I'd recommend that they just pull it out of the parent component altogether, removing the need for any memoization overhead at all. Then, their controls are defined outside of the component in their own function, a function which returns some JSX. A function which returns JSX is a component, so they'd end up with this.

```tsx
const Controls = () => (
  <ControlsBar>
    <PlayPauseButton />
    <SeekControl />
  </ControlsBar>
);

const MyComponent = () => {
  return (
    <VideoPlayer
      renderControls={Controls}
    />
  );
}
```

I'd be inclined to say that that would be an over-optimisation, but the option is there for those who want it.

I appreciate that elsewhere in the codebase this behaviour may be implemented more like this:

```tsx
const MyComponent = () => {
  return (
    <VideoPlayer>
      <VideoPlayer.Controls>
        <ControlsBar>
          <PlayPauseButton />
          <SeekControl />
        </ControlsBar>
      </VideoPlayer.Controls>
    </VideoPlayer>
  );
}
```

And I'm happy to go with that approach instead if you'd prefer. Both approaches achieve the same thing as far as I understand, and are really just a workaround for the fact that React doesn't have a standard way of supporting [slots](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/slot) as you'd see in Vue, Svelte, Web Components, etc.

I chose to use the `renderControls` function instead as that doesn't require iterating over the children in the `VideoPlayer` component and so in theory should be more performant and require less code to achieve the same effect.

> We're surfacing sub-components in the root of the package, some of which could overlap with future component names. E.g. ControlsBar.

More than happy to namespace these to the `VideoPlayer` component, that was an oversight on my part, thanks for the suggestion 🙂

Reply:
Resolving second point ad5863aa264c892bcb1864556e19d512f0913d2b

Reply:
I've modified my approach in 4873e4713bad9c9ce1780008cf36c994d2f7053d, let me know if you prefer it this way

Reply:
This looks good, thanks for explaining the thought process too @joshfarrant 👍 

To address a few of your points...

> defends against future use-cases we might not have considered at this point, eg attaching analytics events to the controls, modifying or adding additional controls, etc.

Yep, agree. In that instance, it would be preferable to use namespaced children like `VideoPlayer.SeekControl` as we can maintain sufficient guardrails internally to the component, whereas an anonymous function would behave more like an escape hatch (e.g. trailingComponent)

> I've not seen this explicit pattern in any other places in our codebase (we do have a couple of renderXyz() functions which return JSX, but none which are taken as props afaict) 

We haven't - to date - made a `renderFoo` slot available on the public API. Though this is more due to naming. Our version of this is leadingComponent and trailingComponent, and we try to be really careful about when/where to use it. This is because we've historically seen a lot of reliance on escape hatches, to the point where we couldn't keep up with all the different variations of components.

Reply:
Just noticed we've dropped the ability to customize the Play icon now that we've moved to a prop based flag. I think that was actually a nice feature, but should be done as a named child with an `icon` prop like we do in `Card`, `Pillar` and `EyebrowBanner`.

E.g.

```jsx
<VideoPlayer>
  <VideoPlayer.PlayPauseControl icon={<svg />} />
</VideoPlayer>
```

Reply:
Previously I had the `renderPlayOverlay` on the `VideoPlayer` which was just used for the full-size play icon while the video was paused. 

I'd suggest that this — along with per-control icon customisation as you mentioned there — be brought in at a later point